### PR TITLE
Cameron/update/session management

### LINKED ISSUE #198: Validate login credentials
#176

### PR DESCRIPTION
**Pull request by camja014**
Alright, so like half of this PR is test related files for the login manager. Since there were like zero server tests before. 

- I added support and classes for Cucumber (BDD test framework I rather like after using it)
- Login manager implementation
- Re-implemented some parts of JungleDB in hopes that I could easily plug-in an in-memory database for testing purposes (still working on that part). It uses a pretty slick [connection pooling library](https://github.com/brettwooldridge/HikariCP) now!
- Moved LoginHandler to a new sub-package

Again, sorry for all the changes.

closes #198

### CODE REVIEW

File: server/pom.xml [hunk 1]
```diff
@@ -36,6 +36,39 @@
             <version>42.1.4</version>
         </dependency>
 
+        <dependency>
```
Comment (camja014) [inline server/pom.xml:39]: Just adding dependencies. Ignore.

File: production.properties [hunk 2]
```diff
@@ -1,7 +1,8 @@
...
-database-user=jungle
-database-password=45v*b1*kGZPo5TR%D!9@caL0%rIjjpBZ
-database-port=5432
\ No newline at end of file
+dataSourceClassName=org.postgresql.ds.PGSimpleDataSource
+dataSource.databaseName=jungle
+dataSource.user=jungle
+dataSource.password=45v*b1*kGZPo5TR%D!9@caL0%rIjjpBZ
+dataSource.serverName=165.227.9.116
+dataSource.portNumber=5432
```
Comment (camja014) [inline production.properties:8]: The Hakari connection pooler uses properties in this format.

File: server/src/main/java/edu/colostate/cs/cs414/chesshireCoders/jungleServer/dataAccessObjects/LoginDAO.java [hunk 3]
```diff
@@ -123,8 +123,8 @@ public void delete(Login login) throws SQLException {
     private Login readLogin(ResultSet rs) throws SQLException {
         Login login = new Login();
         login.setUserID(rs.getInt("UserID"));
-        login.setEmail(rs.getString("Email"));
-        login.setHashedPass(rs.getString("HashedPass"));
+        login.setEmail(rs.getString("Email").trim());
+        login.setHashedPass(rs.getString("HashedPass").trim());
         return login;
```
Comment (camja014) [inline server/src/main/java/edu/colostate/cs/cs414/chesshireCoders/jungleServer/dataAccessObjects/LoginDAO.java:128]: Adding whitespace trimming.

File: server/src/main/java/edu/colostate/cs/cs414/chesshireCoders/jungleServer/handlers/LoginHandler.java [hunk 4]
```diff
@@ -1,75 +0,0 @@
-package edu.colostate.cs.cs414.chesshireCoders.jungleServer.handlers;
-
```
Comment (camja014) [inline server/src/main/java/edu/colostate/cs/cs414/chesshireCoders/jungleServer/handlers/LoginHandler.java:2]: This was just moved

File: server/src/main/java/edu/colostate/cs/cs414/chesshireCoders/jungleServer/handlers/SessionHandler.java [hunk 5]
```diff
@@ -1,80 +0,0 @@
-package edu.colostate.cs.cs414.chesshireCoders.jungleServer.handlers;
```
Comment (camja014) [inline server/src/main/java/edu/colostate/cs/cs414/chesshireCoders/jungleServer/handlers/SessionHandler.java:1]: Also moved.

File: server/src/main/java/edu/colostate/cs/cs414/chesshireCoders/jungleServer/handlers/sessionHandlers/LoginHandler.java [hunk 6]
```diff
@@ -0,0 +1,50 @@
+package edu.colostate.cs.cs414.chesshireCoders.jungleServer.handlers.sessionHandlers;
```
Comment (camja014) [inline server/src/main/java/edu/colostate/cs/cs414/chesshireCoders/jungleServer/handlers/sessionHandlers/LoginHandler.java:1]: Where it was moved to...

File: server/src/main/java/edu/colostate/cs/cs414/chesshireCoders/jungleServer/server/JungleDB.java [hunk 7]
```diff
@@ -45,4 +41,9 @@ public Connection getConnection() throws SQLException {
             throw new SQLException("Data source is not initialized.");
         }
     }
+
+    private void initDataSource(Properties properties) {
+        HikariConfig config = new HikariConfig(properties);
+        dataSource = new HikariDataSource(config);
+    }
```
Comment (camja014) [inline server/src/main/java/edu/colostate/cs/cs414/chesshireCoders/jungleServer/server/JungleDB.java:48]: Updated to use the HakariCP sql connection pooler. It's a lot simpler to use, and it makes databases a lot more pluggable. The hope was I could leverage this to setup an entirely in-memory DB for running tests. Still a WIP.

File: server/src/main/java/edu/colostate/cs/cs414/chesshireCoders/jungleServer/server/JungleServer.java [hunk 8]
```diff
@@ -106,30 +107,6 @@ public void sendToTCP(int connectionID, Object o) {
 
     @Override
     public Connection newConnection() {
-        return new SessionConnection();
-    }
-
-    public class SessionConnection extends Connection {
```
Comment (camja014) [inline server/src/main/java/edu/colostate/cs/cs414/chesshireCoders/jungleServer/server/JungleServer.java:112]: This was moved to its own file.

**Review by amgunden — APPROVED** [review]

**Pull request merged**